compiler/rustc_parse/src/parser/mod.rs

```diff
@@ -1249,7 +1264,14 @@ impl<'a> Parser<'a> {
             append_unglued_token: self.token_cursor.append_unglued_token.clone(),
         };
 
-        let mut ret = f(self)?;
+        let old_capturing = std::mem::replace(&mut self.capturing, true);
```

Comment:
One more thing - looks like we are never resetting `capturing` to false.
So if we enabled it once when collecting tokens for a module:
```rust
#[my_attr]
mod m { ... }
```
we will proceed to collecting tokens for e.g. every single expression in that module.
In the original token collection PR that caused prohibitive performance regressions, why don't they happen now?

Why is `capturing` necessary at all?
It wasn't used previously, and conditional token collection was only enabled by attributes.
It it is necessary only for statements, perhaps it can be implemented as a `collect_tokens_trailing_token` parameter rather than global `Parser` state.

Reply:
`capturing` is initialized to `false` when a `Parser` is created, and gets reset to its original value below.
The performance regressions are fixed by bailing out if we already have tokens, without creating a `LazyTokenStream`.

Without `capturing`, we run into problems when capturing a nontetminal that has attributes. For example,
`my_mac!(#[allow(dead_code)] 25)` will cause us to capture a `TokenStream` that includes `#[allow(dead_code)]`. The expression parsing code will skip collecting tokens, since we only have butin attributes being applied. The outer `collect_tokens` call will then apy it's `TokenStream` to the expression, which includes the attributes.

We could also fix that by unconditionally collecting tokens when we see any attributes. However, I need the `capturing` flag for the rest of the token-based attribute handling PR, so my next PR would just re-add this flag.

Reply:
I misread your module attribute example - that does look like it could be a problem. I'll look at addressing that.